File: code/R/chapter06_planning.R
# Chapter 6: power, sensitivity, and sample-size planning.
# Run from the project root.
#
# Required packages:
# install.packages(c("dplyr", "ggplot2"))

library(dplyr)
library(ggplot2)

scenarios <- read.csv(
  "data/chapter06_planning_scenarios.csv",
  stringsAsFactors = FALSE
)

# Two independent means.
mean_plan <- power.t.test(
  delta = 5,
  sd = 12,
  sig.level = 0.05,
  power = 0.80,
  type = "two.sample",
  alternative = "two.sided"
)

analysis_n_per_group <- ceiling(mean_plan$n)
recruit_n_per_group <- ceiling(
  analysis_n_per_group / (1 - 0.15)
)

print(mean_plan)
print(c(
  analysis_n_per_group = analysis_n_per_group,
  recruit_n_per_group = recruit_n_per_group
))

# Two independent proportions.
proportion_plan <- power.prop.test(
  p1 = 0.60,
  p2 = 0.75,
  sig.level = 0.05,
  power = 0.80,
  alternative = "two.sided"
)

print(proportion_plan)

# Precision-based planning for one mean.
sigma <- 10
half_width <- 2
z_critical <- qnorm(0.975)

n_precision <- ceiling(
  (z_critical * sigma / half_width)^2
)

print(c(
  precision_analysis_n = n_precision,
  precision_recruit_n = ceiling(
    n_precision / 0.90
  )
))

# Cluster design effect.
cluster_size <- 20
icc <- 0.05

design_effect <- 1 + (cluster_size - 1) * icc

print(c(
  design_effect = design_effect,
  individual_equivalent_total = ceiling(
    2 * analysis_n_per_group * design_effect
  )
))

# Approximate ANCOVA efficiency.
baseline_correlation <- 0.60
variance_factor <- 1 - baseline_correlation^2

print(c(
  variance_factor = variance_factor,
  adjusted_n_per_group = ceiling(
    analysis_n_per_group * variance_factor
  )
))

# Power curve.
differences <- seq(1, 8, by = 0.5)
sample_sizes <- c(40, 60, 80, 100, 120, 160, 200)

power_curve <- expand.grid(
  difference = differences,
  n_per_group = sample_sizes
) |>
  rowwise() |>
  mutate(
    power = power.t.test(
      n = n_per_group,
      delta = difference,
      sd = 12,
      sig.level = 0.05,
      type = "two.sample",
      alternative = "two.sided"
    )$power
  ) |>
  ungroup()

ggplot(
  power_curve,
  aes(
    x = difference,
    y = power,
    group = factor(n_per_group),
    linetype = factor(n_per_group)
  )
) +
  geom_line() +
  geom_hline(yintercept = 0.80) +
  labs(
    x = "True mean difference",
    y = "Power",
    linetype = "n per group",
    title = "Power as a function of effect and sample size"
  )

# Type S and Type M simulation.
set.seed(20260715)

simulate_errors <- function(
  n_per_group,
  true_difference = 3,
  sigma = 12,
  repetitions = 20000
) {
  estimates <- numeric(repetitions)
  significant <- logical(repetitions)

  for (i in seq_len(repetitions)) {
    y1 <- rnorm(
      n_per_group,
      mean = true_difference,
      sd = sigma
    )
    y0 <- rnorm(
      n_per_group,
      mean = 0,
      sd = sigma
    )

    result <- t.test(
      y1,
      y0,
      var.equal = FALSE
    )

    estimates[i] <- mean(y1) - mean(y0)
    significant[i] <- result$p.value < 0.05
  }

  selected <- estimates[significant]

  data.frame(
    n_per_group = n_per_group,
    empirical_power = mean(significant),
    type_s = mean(selected < 0),
    type_m = mean(abs(selected)) /
      abs(true_difference),
    mean_estimate_all = mean(estimates),
    mean_estimate_significant = mean(selected)
  )
}

type_sm <- bind_rows(
  lapply(
    c(25, 50, 100, 200),
    simulate_errors
  )
)

print(type_sm)

File: code/Python/chapter06_planning.py
"""Chapter 6: power, sensitivity, and sample-size planning.

Run from the project root:
    python code/Python/chapter06_planning.py

Required packages:
    pandas, numpy, scipy, statsmodels, matplotlib
"""

from pathlib import Path
import math
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy import stats
from statsmodels.stats.power import TTestIndPower, NormalIndPower
from statsmodels.stats.proportion import proportion_effectsize

SCENARIOS = Path("data/chapter06_planning_scenarios.csv")
scenarios = pd.read_csv(SCENARIOS)

# Two independent means.
delta = 5.0
sigma = 12.0
alpha = 0.05
target_power = 0.80

effect_size = delta / sigma
mean_power = TTestIndPower()

n_per_group = mean_power.solve_power(
    effect_size=effect_size,
    alpha=alpha,
    power=target_power,
    ratio=1.0,
    alternative="two-sided",
)

analysis_n = math.ceil(n_per_group)
recruit_n = math.ceil(
    analysis_n / (1 - 0.15)
)

print("Two independent means")
print({
    "standardised_effect": effect_size,
    "analysis_n_per_group": analysis_n,
    "recruit_n_per_group": recruit_n,
})

# Two independent proportions.
p_control = 0.60
p_intervention = 0.75

h = proportion_effectsize(
    p_intervention,
    p_control,
)

proportion_power = NormalIndPower()
n_prop = proportion_power.solve_power(
    effect_size=h,
    alpha=0.05,
    power=0.80,
    ratio=1.0,
    alternative="two-sided",
)

print("\nTwo independent proportions")
print({
    "cohen_h": h,
    "analysis_n_per_group": math.ceil(n_prop),
})

# Precision-based planning for one mean.
sigma_mean = 10.0
half_width = 2.0
z_critical = stats.norm.ppf(0.975)

n_precision = math.ceil(
    (
        z_critical
        * sigma_mean
        / half_width
    ) ** 2
)

print("\nPrecision-based planning")
print({
    "analysis_n": n_precision,
    "target_half_width": half_width,
})

# Attrition inflation.
retention = 1 - 0.15
inflated_n = math.ceil(
    analysis_n / retention
)

print("\nAttrition inflation")
print({
    "analysis_n_per_group": analysis_n,
    "retention": retention,
    "recruit_n_per_group": inflated_n,
})

# Cluster design effect.
cluster_size = 20
icc = 0.05

design_effect = (
    1
    + (cluster_size - 1) * icc
)

print("\nCluster design")
print({
    "design_effect": design_effect,
    "individual_equivalent_total": math.ceil(
        2 * analysis_n * design_effect
    ),
})

# Approximate ANCOVA efficiency.
baseline_correlation = 0.60
variance_factor = 1 - baseline_correlation**2

print("\nANCOVA efficiency")
print({
    "variance_factor": variance_factor,
    "approximate_n_per_group": math.ceil(
        analysis_n * variance_factor
    ),
})

# Power curves.
differences = np.arange(1.0, 8.5, 0.5)
sample_sizes = [40, 60, 80, 100, 120, 160, 200]

curve_rows = []
for difference in differences:
    d_value = difference / sigma
    for n_value in sample_sizes:
        power = mean_power.power(
            effect_size=d_value,
            nobs1=n_value,
            alpha=0.05,
            ratio=1.0,
            alternative="two-sided",
        )
        curve_rows.append({
            "difference": difference,
            "n_per_group": n_value,
            "power": power,
        })

curve = pd.DataFrame(curve_rows)

fig, ax = plt.subplots()
for n_value in sample_sizes:
    subset = curve[
        curve["n_per_group"] == n_value
    ]
    ax.plot(
        subset["difference"],
        subset["power"],
        label=f"n={n_value} per group",
    )

ax.axhline(0.80)
ax.set(
    xlabel="True mean difference",
    ylabel="Power",
    title="Power as a function of effect and sample size",
)
ax.legend()
fig.tight_layout()
plt.show()

# Type S and Type M simulation.
rng = np.random.default_rng(20260715)
true_difference = 3.0
repetitions = 20000

for n_value in [25, 50, 100, 200]:
    estimates = []
    significant = []

    for _ in range(repetitions):
        y1 = rng.normal(
            true_difference,
            sigma,
            n_value,
        )
        y0 = rng.normal(
            0,
            sigma,
            n_value,
        )

        test = stats.ttest_ind(
            y1,
            y0,
            equal_var=False,
        )
        estimate = y1.mean() - y0.mean()

        estimates.append(estimate)
        significant.append(test.pvalue < 0.05)

    estimates = np.asarray(estimates)
    significant = np.asarray(significant)
    selected = estimates[significant]

    type_s = np.mean(selected < 0)
    type_m = (
        np.mean(np.abs(selected))
        / abs(true_difference)
    )

    print({
        "n_per_group": n_value,
        "power": significant.mean(),
        "type_s": type_s,
        "type_m": type_m,
    })